## Overview

This notebook is concerned with four types of geometric objects: Platonic solids, Archimedean solids, regular tilings, and semiregular tilings.

All four classes of objects share a common property: vertex-transitivity. That is, every vertex is identical. What differentiates the four classes is two criteria. First, the solid objects must "fold up" and so the sum of the degrees at each vertex is less than 360°. The tilings must lay perfectly flat and thus must have vertex sums of exactly 360°. Second, the "regular" objects (Platonic solids and regular tilings) must not only have identical vertices, but must also be made up of only one type of polygon. That is, they must also be face-transitive.

The four classes can be identified in a vacuum by simply looking at their vertices:
* Platonic solids
  * Vertex sum < 360°
  * Only one type of polygon at each vertex
* Archimedean solids
  * Vertex sum < 360°
  * More than one type of polygon at each vertex
* Regular tiling of the plane
  * Vertex sum == 360°
  * Only one type of polygon at each vertex
* Semiregular tiling of the plane
  * Vertex sum == 360°
  * More than one type of polygon at each vertex
  
### Motivation

It is common to hear phrases such as "the five platonic solids" or "the three regular tilings of the plane" and students of geometry may wonder: **WHY are there only five platonic solids? HOW DO WE KNOW that there are only five?** There have been many explanations (see the README for more details) but I wanted to construct a simple code-based exhaustive search to answer this question.

Now, we can begin looking for them.

According to wikipedia:
 >In geometry, an Archimedean solid is one of 13 convex polyhedra whose faces are regular polygons and whose vertices are all symmetric to each other

How could we possibly find all of them? Well, it's actually not so hard! We can do this almost entirely in 2 dimensions. **Since these polyhedra are vertex transitive, we can just construct one vertex and it will define the whole polyhedron.**

We know that "doing a loop" is 360°. We also know that traversing the boundary of an n-gon brings us facing back in the same direction we started in. These "turns" we make are really just the exterior angles of an n-gon. Therefore, the sum of the exterior angles of a polygon must be 360°.

From Google:
> An exterior angle is the angle formed outside a polygon when one of its sides is extended. The sum of the exterior angles of any convex polygon is 360°.

We can then find the measure of any individual exterior angle by `360 / n`.

And since we know that for any angle, the sum of the interior and exterior angles is 180° (a straight line), we know that a specific interior angle can be found by `180 - (360 / n)`.

In [2]:
# This method gets the interior angle of an n-gon
def get_interior_angle(n):
    return 180 - (360 / n)

In [3]:
# Let's print the value of a single interior angle for all regular polygons with 3-20 sides:
for n in range(3, 21):
    print(f"A polygon with {n} sides has interior angles of {get_interior_angle(n)}°")

A polygon with 3 sides has interior angles of 60.0°
A polygon with 4 sides has interior angles of 90.0°
A polygon with 5 sides has interior angles of 108.0°
A polygon with 6 sides has interior angles of 120.0°
A polygon with 7 sides has interior angles of 128.57142857142856°
A polygon with 8 sides has interior angles of 135.0°
A polygon with 9 sides has interior angles of 140.0°
A polygon with 10 sides has interior angles of 144.0°
A polygon with 11 sides has interior angles of 147.27272727272728°
A polygon with 12 sides has interior angles of 150.0°
A polygon with 13 sides has interior angles of 152.30769230769232°
A polygon with 14 sides has interior angles of 154.28571428571428°
A polygon with 15 sides has interior angles of 156.0°
A polygon with 16 sides has interior angles of 157.5°
A polygon with 17 sides has interior angles of 158.8235294117647°
A polygon with 18 sides has interior angles of 160.0°
A polygon with 19 sides has interior angles of 161.05263157894737°
A polygon with

In [4]:
# This function returns the sum of all the angles of a given vertex.
def get_vertex_sum(vertex):
	vertex_sum = 0
	for shape in vertex:
		vertex_sum += get_interior_angle(shape)
	return vertex_sum

In [5]:
# For example, the vertices of a cube each have 3 squares (4 sides) meeting together
get_vertex_sum([4, 4, 4])

270.0

Let's narrow down our search a bit.


1. Every "corner" or "vertex" must be comprised of at least 3 polygons. This makes intutive sense if you think of corners of real-life polyhedra. The corner of a desk for example is the intersection of 3 planes. You couldn't make a box, or even one corner of a box with only two flat pieces of wood. If only two faces met at a vertex, there would be no "corner" or solid angle, causing the polyhedron to collapse and not form a three-dimensional shape. This can also be seen numerically by looking at the vertex sum of an angle. For further evidence, we can look at the output above to see that no polygon has an internal angle greater than or equal to 180°. Therefore, every vertex in a polyhedra must be made up of at least 3 faces.
2. Every "vertex" must be comprised of at most 6 polygons. This is because this investigation deals only with geometric forms which are composed of regular polygons. Because the smallest regular polygon (the equilateral triangle) has angles of 60°, we can only fit at most 6 of them around a vertex.
3. The internal angles of the polygons at a vertex must sum to at most 360° because the sum of the angles around a point is 360°. If the sum of the interior angles were to be greater than 360°, the polygons would overlap.
4. Each polygon must have at least 3 sides. This is because you can't construct a polygon with nonzero area with less than 3 sides.

In summary, each vertex of a polyhedra must conform to the following rules:
* The vertex must be made up of at least 3 "faces" (polygons)"
* The vertex must be made up of at most 6 faces
* The vertex sum must be at most 360°
* Each polygon must have at least 3 sides

We could limit our search further with an tight upper bound on the number of sides that a polygon could have. But for now, let's just arbirarily say that a polygon can have at most 99 sides.

Let's start with finding all the geometric objects with vertex configurations of 3 faces.

In [8]:
def get_three_meets():
	to_return = []
	for a in range(3, 100):
		for b in range(3, 100):
			for c in range(3, 100):
				possible_config = (a, b, c)
				vertex_sum = get_vertex_sum(possible_config)
				
				if vertex_sum <= 360:
					to_return.append(possible_config)

	return to_return

all_three_meets = get_three_meets()

In [15]:
print(all_three_meets)
print(len(all_three_meets))

[(3, 3, 3), (3, 3, 4), (3, 3, 5), (3, 3, 6), (3, 3, 7), (3, 3, 8), (3, 3, 9), (3, 3, 10), (3, 3, 11), (3, 3, 12), (3, 3, 13), (3, 3, 14), (3, 3, 15), (3, 3, 16), (3, 3, 17), (3, 3, 18), (3, 3, 19), (3, 3, 20), (3, 3, 21), (3, 3, 22), (3, 3, 23), (3, 3, 24), (3, 3, 25), (3, 3, 26), (3, 3, 27), (3, 3, 28), (3, 3, 29), (3, 3, 30), (3, 3, 31), (3, 3, 32), (3, 3, 33), (3, 3, 34), (3, 3, 35), (3, 3, 36), (3, 3, 37), (3, 3, 38), (3, 3, 39), (3, 3, 40), (3, 3, 41), (3, 3, 42), (3, 3, 43), (3, 3, 44), (3, 3, 45), (3, 3, 46), (3, 3, 47), (3, 3, 48), (3, 3, 49), (3, 3, 50), (3, 3, 51), (3, 3, 52), (3, 3, 53), (3, 3, 54), (3, 3, 55), (3, 3, 56), (3, 3, 57), (3, 3, 58), (3, 3, 59), (3, 3, 60), (3, 3, 61), (3, 3, 62), (3, 3, 63), (3, 3, 64), (3, 3, 65), (3, 3, 66), (3, 3, 67), (3, 3, 68), (3, 3, 69), (3, 3, 70), (3, 3, 71), (3, 3, 72), (3, 3, 73), (3, 3, 74), (3, 3, 75), (3, 3, 76), (3, 3, 77), (3, 3, 78), (3, 3, 79), (3, 3, 80), (3, 3, 81), (3, 3, 82), (3, 3, 83), (3, 3, 84), (3, 3, 85), (3, 3, 86)

First of all - I made a huge error in the above code. Running it we get 2,872 solids, but there's a ton of duplicates! For example, we have (99, 4, 3) and (99, 3, 4) counted as separate polygons. Let's run it again with the constraint that a <= b <= c :

# TODO: explain why this constraint doesn't eliminate any cool solids

In [11]:
def get_three_meets_v2():
	to_return = []
	for c in range(3, 100):
		for b in range(3, c+1):
			for a in range(3, b+1):
				possible_config = (a, b, c)
				vertex_sum = get_vertex_sum(possible_config)
				
				if vertex_sum <= 360:
					to_return.append(possible_config)

	return to_return

all_three_meets_v2 = get_three_meets_v2()

In [14]:
print(all_three_meets_v2)
print(len(all_three_meets_v2))

[(3, 3, 3), (3, 3, 4), (3, 4, 4), (4, 4, 4), (3, 3, 5), (3, 4, 5), (4, 4, 5), (3, 5, 5), (4, 5, 5), (5, 5, 5), (3, 3, 6), (3, 4, 6), (4, 4, 6), (3, 5, 6), (4, 5, 6), (5, 5, 6), (3, 6, 6), (4, 6, 6), (5, 6, 6), (6, 6, 6), (3, 3, 7), (3, 4, 7), (4, 4, 7), (3, 5, 7), (4, 5, 7), (5, 5, 7), (3, 6, 7), (4, 6, 7), (5, 6, 7), (3, 7, 7), (4, 7, 7), (3, 3, 8), (3, 4, 8), (4, 4, 8), (3, 5, 8), (4, 5, 8), (5, 5, 8), (3, 6, 8), (4, 6, 8), (3, 7, 8), (4, 7, 8), (3, 8, 8), (4, 8, 8), (3, 3, 9), (3, 4, 9), (4, 4, 9), (3, 5, 9), (4, 5, 9), (5, 5, 9), (3, 6, 9), (4, 6, 9), (3, 7, 9), (4, 7, 9), (3, 8, 9), (3, 9, 9), (3, 3, 10), (3, 4, 10), (4, 4, 10), (3, 5, 10), (4, 5, 10), (5, 5, 10), (3, 6, 10), (4, 6, 10), (3, 7, 10), (3, 8, 10), (3, 9, 10), (3, 10, 10), (3, 3, 11), (3, 4, 11), (4, 4, 11), (3, 5, 11), (4, 5, 11), (3, 6, 11), (4, 6, 11), (3, 7, 11), (3, 8, 11), (3, 9, 11), (3, 10, 11), (3, 11, 11), (3, 3, 12), (3, 4, 12), (4, 4, 12), (3, 5, 12), (4, 5, 12), (3, 6, 12), (4, 6, 12), (3, 7, 12), (3, 8, 

This gives us 587 vertex configurations already! But wait... do all of these configurations actually correspond to possible solids/tilings?

Here is where the first big insight comes in:

**For any vertex configuration (a, b, c) if a is odd, then b = c.**

How do we know this? A great proof can be found at the following link:
https://www.math.harvard.edu/media/AllenLiuTheStarsAboveUsThesis.pdf

I will explain this reasoning here: "Lemma: If a uniform polyhedron has an odd-sided face, that face must be surrounded at each vertex by two faces that are identical to each other."

We know this because if a is odd, then one of the shapes ("A") that meets at the vertex is an odd-sided polygon. Since all of its vertices have the form (a, b, c), each of its sides must border a polygon with b or c sides. These sides must alternate, such that the polygons bordering A must have a number of sides  bcbcbc...

However, since there is an odd number of sides, this means that there will eventually come a vertex that is of form (a, b, b) or (a, b, c). Therefore, this would not be an Archimedian solid, and so: if a is odd, then b = c.

We can re-state this by saying that, for each vertex (a, b, c) one of the following statements must be true:
a is odd, and b = c
b is odd, and a = c
c id odd, and a = b
or a, b, and c are all even.

This means that we have over-counted. Many of these 587 vertex configurations are not actually possible to construct. Let's try counting again with this constraint in mind.

In [16]:
def is_contradiction(a, b, c):
	a_is_odd = a % 2 == 1
	b_is_odd = b % 2 == 1
	c_is_odd = c % 2 == 1

	if a_is_odd and (b != c):
		return True

	if b_is_odd and (a != c):
		return True

	if c_is_odd and (a != b):
		return True

	return False

def filter_out_contradictions(list_of_solids):
	to_return = []
	for a, b, c in list_of_solids:
		if is_contradiction(a, b, c):
			pass
		else:
			to_return.append((a, b, c))

	return to_return


all_three_meets_v3 = filter_out_contradictions(all_three_meets_v2)

In [17]:
print(all_three_meets_v3)
print(len(all_three_meets_v3))

[(3, 3, 3), (3, 4, 4), (4, 4, 4), (4, 4, 5), (5, 5, 5), (4, 4, 6), (3, 6, 6), (4, 6, 6), (5, 6, 6), (6, 6, 6), (4, 4, 7), (4, 4, 8), (4, 6, 8), (3, 8, 8), (4, 8, 8), (4, 4, 9), (4, 4, 10), (4, 6, 10), (3, 10, 10), (4, 4, 11), (4, 4, 12), (4, 6, 12), (3, 12, 12), (4, 4, 13), (4, 4, 14), (4, 4, 15), (4, 4, 16), (4, 4, 17), (4, 4, 18), (4, 4, 19), (4, 4, 20), (4, 4, 21), (4, 4, 22), (4, 4, 23), (4, 4, 24), (4, 4, 25), (4, 4, 26), (4, 4, 27), (4, 4, 28), (4, 4, 29), (4, 4, 30), (4, 4, 31), (4, 4, 32), (4, 4, 33), (4, 4, 34), (4, 4, 35), (4, 4, 36), (4, 4, 37), (4, 4, 38), (4, 4, 39), (4, 4, 40), (4, 4, 41), (4, 4, 42), (4, 4, 43), (4, 4, 44), (4, 4, 45), (4, 4, 46), (4, 4, 47), (4, 4, 48), (4, 4, 49), (4, 4, 50), (4, 4, 51), (4, 4, 52), (4, 4, 53), (4, 4, 54), (4, 4, 55), (4, 4, 56), (4, 4, 57), (4, 4, 58), (4, 4, 59), (4, 4, 60), (4, 4, 61), (4, 4, 62), (4, 4, 63), (4, 4, 64), (4, 4, 65), (4, 4, 66), (4, 4, 67), (4, 4, 68), (4, 4, 69), (4, 4, 70), (4, 4, 71), (4, 4, 72), (4, 4, 73), (4, 4

Looking at these 164 configurations, we notice that many of them belong to a unique family: (4, 4, x). Polyhedra in this family are called "prisms," and there are infinitely many of them. However, there is an exception. The cube (4, 4, 4) is a Platonic solid and so is usually not classified as part of this infinite family. For now, let's filter out the infinite family of prisms.

In [18]:
def is_prism(a, b, c):
    return [a, b, c].count(4) == 2

#def filter_out_infinite_families(list_of_solids):
def filter_out_prisms(list_of_solids):
	to_return = []
	for a, b, c in list_of_solids:
		if is_prism(a, b, c):
			pass
		else:
			to_return.append((a, b, c))

	return to_return


all_three_meets_v4 = filter_out_prisms(all_three_meets_v3)

In [19]:
print(all_three_meets_v4)
print(len(all_three_meets_v4))

[(3, 3, 3), (4, 4, 4), (5, 5, 5), (3, 6, 6), (4, 6, 6), (5, 6, 6), (6, 6, 6), (4, 6, 8), (3, 8, 8), (4, 8, 8), (4, 6, 10), (3, 10, 10), (4, 6, 12), (3, 12, 12)]
14


![Test image](https://en.wikipedia.org/wiki/Truncated_tetrahedron#/media/File:Truncated_tetrahedron.png)

We have finally arrived at all possible vertex configurations of 3 polygons. Let's go through them one-by-one.


Thanks to user [Cyp](https://en.wikipedia.org/wiki/User:Cyp) on Wikipedia for the images.






#### Tetrahedron (3, 3, 3)

![img](Tetrahedron.png)

// write a bit

#### Cube (4, 4, 4)

![img](Cube.png)

// write a bit

#### Dodecahedron (5, 5, 5)

![img](Dodecahedron.png)


#### [Truncated tetrahedron](https://en.wikipedia.org/wiki/Truncated_tetrahedron) (3, 6, 6)
#### [Truncated octahedron](https://en.wikipedia.org/wiki/Truncated_octahedron) (4, 6, 6)
#### [Truncated icosahedron](https://en.wikipedia.org/wiki/Truncated_icosahedron) (5, 6, 6)
#### [Hexagonal tiling](https://en.wikipedia.org/wiki/Hexagonal_tiling) (6, 6, 6)
#### [Truncated cuboctahedron](https://en.wikipedia.org/wiki/Truncated_cuboctahedron) (4, 6, 8)
#### [Truncated cube](https://en.wikipedia.org/wiki/Truncated_cube) (3, 8, 8)
#### [Truncated square tiling](https://en.wikipedia.org/wiki/Truncated_square_tiling) (4, 8, 8)
#### [Truncated icosidodecahedron](https://en.wikipedia.org/wiki/Truncated_icosidodecahedron) (4, 6, 10)
#### [Truncated dodecahedron](https://en.wikipedia.org/wiki/Truncated_dodecahedron) (3, 10, 10)
#### [Truncated trihexagonal tiling](https://en.wikipedia.org/wiki/Truncated_trihexagonal_tiling) (4, 6, 12)
#### [Truncated hexagonal tiling](https://en.wikipedia.org/wiki/Truncated_hexagonal_tiling) (3, 12, 12)


To review:
* (3, 3, 3)		Platonic Solid (#1)
* (4, 4, 4)		Platonic Solid (#2)
* (5, 5, 5)		Platonic Solid (#3)
* (3, 6, 6)		Archimedian Solid (#1)
* (4, 6, 6)		Archimedian Solid (#2)
* (5, 6, 6)		Archimedian Solid (#3)
* (6, 6, 6)		Regular Tiling (#1)
* (4, 6, 8)		Archimedian Solid (#4)
* (3, 8, 8)		Archimedian Solid (#5)
* (4, 8, 8)		Semiregular Tiling (#1)
* (4, 6, 10)	Archimedian Solid (#6)
* (3, 10, 10)	Archimedian Solid (#7)
* (4, 6, 12)	Semiregular Tiling (#2)
* (3, 12, 12)	Semiregular Tiling (#3)

3 platonic solids
7 archimedian solids
1 regular tiling
3 semiregular tilings

Let's move on to the next part, 4-face vertices. Here, our searching logic here needs to be different. Not every vertex can be re-written in a format where it's decreasing or equal.

We have a weird issue. It has to do with symmetries of regular polygons. A triangle has 6 [symmetries](https://en.wikipedia.org/wiki/Dihedral_group) and a square has 8. However, for 3 numbers, there are only 3! = 6 orders to write them in. But for 4 numbers, there are 4! = 24 ways. Therefore, we have an issue. We can't just allow all duplicates, as for any tuple (a, b, c, d), there are 8 valid ways to write it. But if we add the requirement that a <= b <= c <= d, then we will discount completely valid configurations, such as (3, 5, 3, 5), the icosidodecahedron.

The actual solution to this is to write up an explanation before we add the
a <= b <= c rule in the 3-meets logic. Anyways.

In [ ]:
def generate_unique_four_meets():
	to_return = []
	for d in range(3, 100):
		for c in range(3, d+1):
			for b in range(3, d+1):
				for a in range(3, d+1):
					possible_config = (a, b, c, d)
					vertex_sum = get_vertex_sum(possible_config)
				
					if vertex_sum > 360:
						continue

					if (a, b, c, d) in to_return or \
					(b, c, d, a) in to_return or \
					(c, d, a, b) in to_return or \
					(d, a, b, c) in to_return or \
					(d, c, b, a) in to_return or \
					(c, b, a, d) in to_return or \
					(b, a, d, c) in to_return or \
					(a, d, c, b) in to_return:
						continue

					to_return.append(possible_config)

	return to_return


all_four_meets_v1 = generate_unique_four_meets()

In [22]:
print(all_four_meets_v1)
print(len(all_four_meets_v1))

[(3, 3, 3, 3), (3, 3, 3, 4), (4, 3, 3, 4), (3, 4, 3, 4), (4, 4, 3, 4), (4, 4, 4, 4), (3, 3, 3, 5), (4, 3, 3, 5), (5, 3, 3, 5), (3, 4, 3, 5), (4, 4, 3, 5), (3, 5, 3, 5), (4, 3, 4, 5), (3, 3, 3, 6), (4, 3, 3, 6), (5, 3, 3, 6), (6, 3, 3, 6), (3, 4, 3, 6), (4, 4, 3, 6), (3, 5, 3, 6), (3, 6, 3, 6), (4, 3, 4, 6), (3, 3, 3, 7), (4, 3, 3, 7), (5, 3, 3, 7), (3, 4, 3, 7), (3, 5, 3, 7), (3, 3, 3, 8), (4, 3, 3, 8), (3, 4, 3, 8), (3, 3, 3, 9), (4, 3, 3, 9), (3, 4, 3, 9), (3, 3, 3, 10), (4, 3, 3, 10), (3, 4, 3, 10), (3, 3, 3, 11), (4, 3, 3, 11), (3, 4, 3, 11), (3, 3, 3, 12), (4, 3, 3, 12), (3, 4, 3, 12), (3, 3, 3, 13), (3, 3, 3, 14), (3, 3, 3, 15), (3, 3, 3, 16), (3, 3, 3, 17), (3, 3, 3, 18), (3, 3, 3, 19), (3, 3, 3, 20), (3, 3, 3, 21), (3, 3, 3, 22), (3, 3, 3, 23), (3, 3, 3, 24), (3, 3, 3, 25), (3, 3, 3, 26), (3, 3, 3, 27), (3, 3, 3, 28), (3, 3, 3, 29), (3, 3, 3, 30), (3, 3, 3, 31), (3, 3, 3, 32), (3, 3, 3, 33), (3, 3, 3, 34), (3, 3, 3, 35), (3, 3, 3, 36), (3, 3, 3, 37), (3, 3, 3, 38), (3, 3, 3, 39

In [23]:
def is_antiprism(vertex):
    return vertex.count(3) == 3

def filter_out_antiprisms(list_of_solids):
	to_return = []
	for config in list_of_solids:
		if is_antiprism(config):
			pass
		else:
			to_return.append(config)

	return to_return


all_four_meets_v2 = filter_out_antiprisms(all_four_meets_v1)

In [24]:
print(all_four_meets_v2)
print(len(all_four_meets_v2))

[(3, 3, 3, 3), (4, 3, 3, 4), (3, 4, 3, 4), (4, 4, 3, 4), (4, 4, 4, 4), (4, 3, 3, 5), (5, 3, 3, 5), (3, 4, 3, 5), (4, 4, 3, 5), (3, 5, 3, 5), (4, 3, 4, 5), (4, 3, 3, 6), (5, 3, 3, 6), (6, 3, 3, 6), (3, 4, 3, 6), (4, 4, 3, 6), (3, 5, 3, 6), (3, 6, 3, 6), (4, 3, 4, 6), (4, 3, 3, 7), (5, 3, 3, 7), (3, 4, 3, 7), (3, 5, 3, 7), (4, 3, 3, 8), (3, 4, 3, 8), (4, 3, 3, 9), (3, 4, 3, 9), (4, 3, 3, 10), (3, 4, 3, 10), (4, 3, 3, 11), (3, 4, 3, 11), (4, 3, 3, 12), (3, 4, 3, 12)]
33


# Now it's time for epic Kepler insight number 2:
# "[There cannot be a] solid angle surrounded by four faces with at least one
# 3-gon and a b-gon opposite of it where a ≠ c.”

In [26]:
def is_lemma2_contradiction(vertex):
    # First, we find the 3-gon
    triangle_indices = []
    for i in range(len(vertex)):
    	if vertex[i] == 3:
    		triangle_indices.append(i)

    for i in triangle_indices:
    	a_index = (i + 1) % 4
    	c_index = (i + 3) % 4
    	if vertex[a_index] != vertex[c_index]:
    		return True

    return False

def filter_out_lemma2_cases(list_of_solids):
	to_return = []
	for config in list_of_solids:
		if is_lemma2_contradiction(config):
			pass
		else:
			to_return.append(config)

	return to_return


all_four_meets_v3 = filter_out_lemma2_cases(all_four_meets_v2)

In [28]:
print(all_four_meets_v3)
print(len(all_four_meets_v3))

[(3, 3, 3, 3), (3, 4, 3, 4), (4, 4, 3, 4), (4, 4, 4, 4), (3, 5, 3, 5), (4, 3, 4, 5), (3, 6, 3, 6), (4, 3, 4, 6)]
8


We have finally arrived at all possible vertex configurations of 4 polygons. Let's go through them one-by-one

blah blah

To review:

#### (3, 3, 3, 3) Platonic Solid (#4)
#### (3, 4, 3, 4) Archimedian Solid (#6)
#### (4, 4, 3, 4) Archimedian Solid (#9)
#### (4, 4, 4, 4) Regular Tiling (#2)
#### (3, 5, 3, 5) Archimedian Solid (#10)
#### (4, 3, 4, 5) Archimedian Solid (#11)
#### (3, 6, 3, 6) Semiregular Tiling (#4)
#### (4, 3, 4, 6) Semiregular Tiling (#5)


1 platonic solid
4 archimedian solids
1 regular tiling
2 semi-regular tilings


We've found 11 of the 13 archimedian solids now. We're in the home stretch! Let's repeat this with the case of 5-face vertices.

For this one, let's go about this in a clever way. We know that the first 4 polygons are going to each contribute at least 60° to the vertex sum. This means that the 5th and final polygon must contribute at most 360° - (4 * 60°) = 360° - 240° = 120°. Therefore, it is clear that even the largest polygon in the largest case has at most 120, which is a hexagon. Let's not waste time searching for polygons with more than 6 sides then!

In [29]:
def generate_unique_five_meets():
	to_return = []
	for e in range(3,7):
		for d in range(3, e+1):
			for c in range(3, e+1):
				for b in range(3, e+1):
					for a in range(3, e+1):
						possible_config = (a, b, c, d, e)
						vertex_sum = get_vertex_sum(possible_config)
					
						if vertex_sum > 360:
							continue

						if (a, b, c, d, e) in to_return or \
						(b, c, d, e, a) in to_return or \
						(c, d, e, a, b) in to_return or \
						(d, e, a, b, c) in to_return or \
						(e, a, b, c, d) in to_return or \
						(e, d, c, b, a) in to_return or \
						(d, c, b, a, e) in to_return or \
						(c, b, a, e, d) in to_return or \
						(b, a, e, d, c) in to_return or \
						(a, e, d, c, b) in to_return:
							continue

						to_return.append(possible_config)

	return to_return


all_five_meets_v1 = generate_unique_five_meets()

In [30]:
print(len(all_five_meets_v1))
for possible_config in all_five_meets_v1:
	print(possible_config, get_vertex_sum(possible_config))

6
(3, 3, 3, 3, 3) 300.0
(3, 3, 3, 3, 4) 330.0
(4, 3, 3, 3, 4) 360.0
(3, 4, 3, 3, 4) 360.0
(3, 3, 3, 3, 5) 348.0
(3, 3, 3, 3, 6) 360.0


# This generates 6 vertex configurations of 5 faces, all of which are valid:

# Let's go through them one-by-one

# blah blah

# To review:

# (3, 3, 3, 3, 3) 300.0 Platonic Solid (#5)
# (3, 3, 3, 3, 4) 330.0 Archimedian Solid (#12)
# (4, 3, 3, 3, 4) 360.0 Semiregular Tiling (#6)
# (3, 4, 3, 3, 4) 360.0 Semiregular Tiling (#7)
# (3, 3, 3, 3, 5) 348.0 Archimedian Solid (#13)
# (3, 3, 3, 3, 6) 360.0 Semiregular Tiling (#8)

# Finally, we can consider the only way that 6 regular polygons can meet at a
# single vertex:

# (3, 3, 3, 3, 3, 3) 360.0 Regular Tiling (#3)





# Now, we consider Kepler’s second lemma.
# Lemma 2: A polyhedron that has the same pattern of polygon faces at every vertex
# cannot have these types of solid angles:
# i) A solid angle surrounded by three faces—a-gon, b-gon, and c-gon—where a is
# odd and b ≠ c. See Figure 2.1. 
# 6
# ii) A solid angle surrounded by four faces with at least one 3-gon and a b-gon
# opposite of it where a ≠ c.” (Cromwell 159). See Figure 2.2
